File: jan17/cowcode.java
// USACO 2017 Silver Contest
// Problem 3: Secret Cow Code

import java.util.*;
import java.io.*;

public class cowcode {

	public static void main(String[] args) throws Exception {
		
		/**
		 * 
		 * This appears to be a string processing problem.
		 * It would be too inefficient to simulate the duplication, but it acts in a trend
		 * 
		 * The last char of the first substring is equal to the first char of the last substring
		 * The rest of the first substring goes after the first char and the index shifts by 1
		 * 
		 * **/
		
		Scanner sc = new Scanner(new FileInputStream(new File("cowcode.in")));
		PrintWriter pw = new PrintWriter(new OutputStreamWriter(new FileOutputStream(new File("cowcode.out"))));
		String str = sc.next();
		long ind = sc.nextLong();
		sc.close();
		pw.println(get(str, ind-1));
		pw.flush();
		pw.close();
		
	}
	
	public static char get(String str, long ind) {
		if(ind < str.length()) {
			return str.charAt((int)ind);
		}
		long l = str.length();
		while(2*l <= ind) {
			l *= 2;
		}
		if(l == ind) {
			return get(str, l-1);
		}
		return get(str, ind-l-1);
	}
	
}

File: jan17/Cowcode.cpp
// January 2017 Silver Contest
// 3. Secret Cow Code

#include <iostream>
#include <string>
#include <stdio.h>
using namespace std;

char get(string s, long long i){
    
    if(i < s.length()){
        return s.at((int)i);
    } // if i is in the string
    long long l = s.length();
    while(2 * l <= i) // while twice the length is less than the index
        l *= 2; // this brings it to said length
    if (l == i) // if length is equal to index
        return get(s, l-1);
    else return get(s, i-l-1);
    
}

int main(){
    
    freopen("cowcode.in", "r", stdin);
    freopen("cowcode.out", "w", stdout);
    
    string s;
    cin >> s;
    long long i;
    cin >> i;
    cout << get(s, i-1) << endl;
    
}
